libs/utils/src/http/endpoint.rs

```diff
@@ -328,6 +336,120 @@ pub async fn prometheus_metrics_handler(_req: Request<Body>) -> Result<Response<
+        .as_ref()
+        .ok_or(ApiError::InternalServerError(anyhow!(
+            "heap profiling not enabled"
+        )))?
+        .lock()
+        .await;
+    if !prof_ctl.activated() {
+        return Err(ApiError::InternalServerError(anyhow!(
+            "heap profiling not enabled"
+        )));
+    }
+
+    // Take and return the profile.
+    match format {
+        Format::Jemalloc => {
+            let file = tokio::task::spawn_blocking(move || prof_ctl.dump())
+                .await
+                .map_err(|join_err| ApiError::InternalServerError(join_err.into()))?
+                .map_err(ApiError::InternalServerError)?;
+            let stream = ReaderStream::new(tokio::fs::File::from_std(file));
+            Response::builder()
+                .status(200)
+                .header(CONTENT_TYPE, "application/octet-stream")
+                .header(CONTENT_DISPOSITION, "attachment; filename=\"heap.dump\"")
+                .body(Body::wrap_stream(stream))
+                .map_err(|err| ApiError::InternalServerError(err.into()))
+        }
+
+        Format::Pprof => {
+            let data = tokio::task::spawn_blocking(move || prof_ctl.dump_pprof())
+                .await
+                .map_err(|join_err| ApiError::InternalServerError(join_err.into()))?
+                .map_err(ApiError::InternalServerError)?;
+            Response::builder()
+                .status(200)
+                .header(CONTENT_TYPE, "application/octet-stream")
+                .header(CONTENT_DISPOSITION, "attachment; filename=\"heap.pb\"")
+                .body(Body::from(data))
+                .map_err(|err| ApiError::InternalServerError(err.into()))
+        }
+
+        Format::Svg => {
+            let tempfile = tokio::task::spawn_blocking(move || {
+                let mut file = prof_ctl.dump()?;
+                // Write the profile to a tempfile we can get the path to. If dump() returned the
+                // NamedTempFile instead of a File then this wouldn't be necessary.
+                let mut tempfile = NamedUtf8TempFile::new()?;
+                std::io::copy(&mut file, &mut tempfile)?;
+                Ok(tempfile)
+            })
+            .await
+            .map_err(|join_err| ApiError::InternalServerError(join_err.into()))?
+            .map_err(ApiError::InternalServerError)?;
+            let svg = jeprof_svg(tempfile.path())
+                .await
+                .map_err(ApiError::InternalServerError)?;
+            Response::builder()
+                .status(200)
+                .header(CONTENT_TYPE, "image/svg+xml")
+                .body(Body::from(svg))
+                .map_err(|err| ApiError::InternalServerError(err.into()))
+        }
+    }
+}
+
+/// Generates an SVG for the given jemalloc profile file. Requires Perl installed.
+///
+/// TODO: jeprof is really slow (~20 seconds) because of addr2line.
+pub async fn jeprof_svg(path: &Utf8Path) -> anyhow::Result<Vec<u8>> {
```

Comment:
Should we use `&Path` here so we can avoid the copy above?

Reply:
You mean the `std::io::copy`? That's not because of `&Path`, it's because the file returned from `dump()` has already been deleted from the file system so we can't get a path to it. I clarified the comment there.